### PR TITLE
Multi-index, merging, docs, etc.

### PR DESCRIPTION
**Pull request by virtualritz**

### CODE REVIEW

**Review by Twinklebear — COMMENTED** [review]
Overall I think this looks really good, I put some small comments in that either of us could handle. I think the only real change requested is adding validation of the `LoadOptions`. I should be able to pick that up after merging, or if you've got time that'd also be great. Same for putting in some new tests of the other load options for CI

**Comment by Twinklebear** [comment]
It looks like for CI we should also take out the `#[feature(test)]` use, is that still needed for testing? Or that's been stabilized now?

File: src/lib.rs [hunk 1]
```diff
@@ -160,111 +273,253 @@ use std::str::{FromStr, SplitWhitespace};
...
+#[derive(Debug, Clone, Copy)]
+pub struct LoadOptions {
+    /// Merge identical positions.
+    ///
+    /// This is usually what you want if you intend to use the mesh in an
+    /// *offline rendering* context or to do further processing with
+    /// *topological operators*.
+    ///
+    /// * This flag has *no effect* if
+    ///   [`single_index`](LoadOptions::single_index) is set!
```
[the remark this answers is not in the document]
Comment (virtualritz) [reply src/lib.rs:360]: Ok, added this.

File: src/lib.rs [hunk 2]
```diff
@@ -498,78 +766,654 @@ fn add_vertex(
...
+
+/// Load the various objects specified in the `OBJ` file and any associated
+/// `MTL` file.
+///
+/// Returns a pair of `Vec`s containing the loaded models and materials from the
+/// file.
+///
+/// # Arguments
+///
+/// * `create_multiple_indices` - Create additional indices for normals and
```
[the remark this answers is not in the document]
Comment (virtualritz) [reply src/lib.rs:1400]: Fixed.

File: src/lib.rs [hunk 3]
```diff
@@ -641,23 +1492,29 @@ where
...
 ///         _ => unreachable!(),
-///     }
-/// });
+///     },
+/// );
 /// ```
 pub fn load_obj_buf<B, ML>(
     reader: &mut B,
-    triangulate_faces: bool,
+    load_options: &LoadOptions,
```
[the remark this answers is not in the document]
Comment (virtualritz) [reply src/lib.rs:1517]: Done.

File: tests/lib.rs [hunk 4]
```diff
@@ -1,17 +1,22 @@
...
+    fs::File,
+    io::{BufReader, Cursor},
+};
 
 const CORNELL_BOX_OBJ: &'static str = include_str!("../cornell_box.obj");
 const CORNELL_BOX_MTL1: &'static str = include_str!("../cornell_box.mtl");
 const CORNELL_BOX_MTL2: &'static str = include_str!("../cornell_box2.mtl");
 
 #[test]
 fn simple_triangle() {
```
[the remark this answers is not in the document]
Comment (virtualritz) [reply tests/lib.rs:12]: Yeah, need to find some time for that.

File: README.md [hunk 5]
```diff
@@ -1,63 +1,116 @@
...
+use tobj;
 
 fn main() {
-    let obj_file = env::args()
+    let obj_file = std::env::args()
         .skip(1)
         .next()
         .expect("A .obj file to print is required");
+
     let (models, materials) = tobj::load_obj(&obj_file, false).expect("Failed to load file");
```
[the remark this answers is not in the document]
Comment (virtualritz) [reply README.md:98]: So I am confused. You want `single_index` but no `triangulate`? For GPU/realtime use they usually would both be on.
And that is for the `lib.rs` example, right?

While in `README.md` you want to show the different indices, i.e. use `LoadOptions::default()`?

[the remark this answers is not in the document]
Comment (virtualritz) [reply README.md]: I did just that. Updated `print_mesh.rs` to support multi index files and made the sample in `lib.rs` use `single_index` *and* `triangulate`.

File: tests/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (virtualritz) [reply tests/lib.rs:14]: I moved `tests/lib.rs` to `src/tests.rs`. I didn't add any tests for other options yet though.

**Comment by virtualritz** [comment]
> It looks like for CI we should also take out the #[feature(test)] use, is that still needed for testing?

I cleaned that up.

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (Twinklebear) [reply README.md]: Ah yes with triangulate as well, I forgot about that option

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Twinklebear) [reply src/lib.rs]: Looks good!

**Comment by Twinklebear** [comment]
The readme change we can probably drop from this PR. For the PR, did you want to add tests as part of it? Or leave that for later? If you wanted to wait on the tests I think this is ready to merge

**Comment by virtualritz** [comment]
Yes, I only added that section since I got some messages asking me about it because of the Rust game newsletter which mentions `tobj` under my `opensubdiv-petite` contrib.

I do not have time to add tests atm. I.e. I'd release to get it out in the wild – fastest way to find issues if I overlooked something. 

That said: I am using all of the new features in my code; it is just a bit of work to isolate that into small tests. I'll get to it eventually.

**Pull request merged**